React/Modules/RCTDevSettings.mm

```diff
@@ -0,0 +1,572 @@
+ * of patent rights can be found in the PATENTS file in the same directory.
+ */
+
+#import "RCTDevSettings.h"
+
+#import <objc/runtime.h>
+#import <JavaScriptCore/JavaScriptCore.h>
+#import <jschelpers/JavaScriptCore.h>
+
+#import "JSCSamplingProfiler.h"
+
+#import "RCTBridge+Private.h"
+#import "RCTBridgeModule.h"
+#import "RCTDevMenu.h"
+#import "RCTEventDispatcher.h"
+#import "RCTLog.h"
+#import "RCTProfile.h"
+#import "RCTUtils.h"
+#import "RCTWebSocketObserverProtocol.h"
+
+NSString *const kRCTDevSettingProfilingEnabled = @"profilingEnabled";
+NSString *const kRCTDevSettingHotLoadingEnabled = @"hotLoadingEnabled";
+NSString *const kRCTDevSettingLiveReloadEnabled = @"liveReloadEnabled";
+NSString *const kRCTDevSettingIsInspectorShown = @"showInspector";
+NSString *const kRCTDevSettingIsDebuggingRemotely = @"isDebuggingRemotely";
+NSString *const kRCTDevSettingWebsocketExecutorName = @"websocket-executor-name";
+NSString *const kRCTDevSettingExecutorOverrideClass = @"executor-override";
+NSString *const kRCTDevSettingShakeToShowDevMenu = @"shakeToShow";
+NSString *const kRCTDevSettingIsPerfMonitorShown = @"RCTPerfMonitorKey";
+NSString *const kRCTDevSettingIsJSCProfilingEnabled = @"RCTJSCProfilerEnabled";
+NSString *const kRCTDevSettingStartSamplingProfilerOnLaunch = @"startSamplingProfilerOnLaunch";
+
+NSString *const kRCTDevSettingsUserDefaultsKey = @"RCTDevMenu";
+
+#if RCT_DEV
+
+@interface RCTDevSettingsUserDefaultsDataSource : NSObject <RCTDevSettingsDataSource>
+
+@property (nonatomic, strong) NSMutableDictionary *settings;
+@property (nonatomic, strong) NSMutableSet *keysChanged;
+
+@end
+
+@implementation RCTDevSettingsUserDefaultsDataSource
+
+- (instancetype)init
+{
+  return [self initWithDefaultValues:nil];
+}
+
+- (instancetype)initWithDefaultValues:(NSDictionary *)defaultValues
+{
+  if (self = [super init]) {
+    if (defaultValues) {
+      [self _reloadWithDefaults:defaultValues];
+    }
+  }
+  return self;
+}
+
+- (void)dealloc
+{
+  [[NSNotificationCenter defaultCenter] removeObserver:self];
+}
+
+- (void)updateSettingWithValue:(id)value forKey:(NSString *)key
+{
+  RCTAssert((key != nil), @"%@", [NSString stringWithFormat:@"%@: Tried to update nil key", [self class]]);
+
+  id currentValue = _settings[key];
+  if (currentValue == value || [currentValue isEqual:value]) {
+    return;
+  }
+  if (value) {
+    _settings[key] = value;
+  } else {
+    [_settings removeObjectForKey:key];
+  }
+  if (!_keysChanged) {
+    _keysChanged = [NSMutableSet set];
+  }
+  [_keysChanged addObject:key];
+  [[NSUserDefaults standardUserDefaults] setObject:_settings forKey:kRCTDevSettingsUserDefaultsKey];
+  [[NSUserDefaults standardUserDefaults] synchronize];
+}
+
+- (id)settingForKey:(NSString *)key
+{
+  return _settings[key];
+}
+
+- (void)_reloadWithDefaults:(NSDictionary *)defaultValues
+{
+  NSDictionary *existingSettings = [[NSUserDefaults standardUserDefaults] objectForKey:kRCTDevSettingsUserDefaultsKey];
```

Comment:
Perf tip: grab the `[NSUserDefaults standardUserDefaults]` object once and reference that instead of calling the `standardUserDefaults` method multiple times.